apps/remote_control/lib/lexical/remote_control/search/store/mnesia/connection.ex

```diff
@@ -0,0 +1,206 @@
+  alias Lexical.RemoteControl
+  alias Lexical.RemoteControl
+  alias Lexical.RemoteControl.Search.Store.Mnesia.Schema
+  alias Lexical.RemoteControl.Search.Store.Mnesia.State
+
+  @leader_name :mnesia_leader
+
+  def connect(%State{} = state) do
+    new_state =
+      case :global.register_name(@leader_name, self()) do
+        :yes ->
+          start_port(state)
+
+        :no ->
+          leader_pid = :global.whereis_name(@leader_name)
+          Process.monitor(leader_pid)
+          %State{state | leader?: false, leader_pid: leader_pid}
+      end
+
+    {:connect_to_node, new_state}
+  end
+
+  def connect_to_node(%State{} = state) do
+    with true <- mnesia_running?(state.project),
+         true <- connect_to_mnesia(state.project) do
+      new_state = %State{state | mnesia_node: mnesia_node_name(state.project)}
+      initialize_mnesia(new_state)
+    else
+      _ ->
+        {:connect_to_node, state}
+    end
+  end
+
+  def on_nodedown(%State{leader?: true, mnesia_node: mnesia_node} = state, mnesia_node) do
+    {:ok, port, ref} = initialize_mnesia_node(state.project)
+    {:connect_to_node, %State{state | port: port, port_ref: ref}}
+  end
+
+  def on_nodedown(%State{} = state, _) do
+    {:connect_to_node, state}
+  end
+
+  def on_port_closed(%State{leader?: true, port_ref: port_ref} = state, port_ref) do
+    {:ok, port, ref} = initialize_mnesia_node(state.project)
+    new_state = %State{state | port: port, port_ref: ref}
+    {:connect_to_node, new_state}
+  end
+
+  def on_port_closed(%State{} = state, _) do
+    {:connect_to_node, state}
+  end
+
+  defp start_port(%State{} = state) do
+    caller = self()
+
+    with_leader_lock(:start_port, fn ->
+      with :ok <- ensure_leader(caller),
+           {:ok, port, ref} <- initialize_mnesia_node(state.project) do
+        %State{state | port: port, port_ref: ref, leader?: true}
+      end
+    end)
+  end
+
+  defp initialize_mnesia(%State{leader?: true} = state) do
+    # Note: It would be great to guarantee that the leader comes through here first.
+    # I don't know if this will do it or not.
+    with_leader_lock(:initialize, fn ->
+      with :ok <- Schema.ensure_schema(state),
+           :ok <- start_remote_mnesia(state),
+           :ok <- Schema.create_table(state) do
+        join_mnesia_as_ram_copy(state)
+        Schema.clean_old_ram_copies(state)
+        {:connected, state}
+      end
+    end)
+  end
+
+  defp initialize_mnesia(%State{} = state) do
+    with_leader_lock(:initialize, fn ->
+      :ok = join_mnesia_as_ram_copy(state)
+      {:connected, state}
+    end)
+  end
+
+  defp initialize_mnesia_node(%Project{} = project) do
+    path_args = Enum.flat_map(:code.get_path(), fn path -> ["-pa", "\"#{path}\""] end)
+
+    port_args = [
+      "--name",
+      to_string(mnesia_node_name(project)),
+      "--cookie",
+      Node.get_cookie(),
+      "-e",
+      ~s[Node.connect(#{inspect(Node.self())})],
+      "--no-halt"
+      | path_args
+    ]
```

Comment:
I noticed that you're calling `to_string(mnesia_node_name(project))`, but we're not stringifying the result of `Node.get_cookie()`, which returns the cookie as an atom. IMO, let's not worry about stringifying at all here, and let's just map `to_string` over the args in `RemoteControl.Port.open_elixir/2`.

```suggestion
    self = inspect(Node.self())

    port_args = [
      "--name",
      mnesia_node_name(project),
      "--cookie",
      Node.get_cookie(),
      "-e",
      "Node.connect(#{self})",
      "--no-halt"
      | path_args
    ]
```